keps/sig-network/2104-reworking-kube-proxy-architecture/README.md

```diff
@@ -0,0 +1,1003 @@
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+At the beginning, `kube-proxy` was designed to handle the translation of Service objects to OS-level
+resources.  Implementations have been userland, then iptables, and now ipvs. With the growth of the
+Kubernetes project, more implementations came to life, for instance with eBPF, and often in relation
+to other goals (Calico to manage the network overlay, Cilium to manage app-level security, metallb
+to provide an external LB for bare-metal clusters, etc).
+
+Along this cambrian explosion of third-party software, the Service object itself received new
+concepts to improve the abstraction, for instance to express topology. Thus, third-party
+implementation are expected to update and become more complex over time, even if their core doesn't
+change (ie, the eBPF translation layer is not affected).
+
+This KEP is born from the conviction that more decoupling of the Service object and the actual
+implementations is required, by introducing an intermediate, node-level abstraction provider. This
+abstraction is expected to be the result of applying Kubernetes' `Service` semantics and business
+logic to a simpler, more stable API.
+
+## Motivation
+
+<!--
+This section is for explicitly listing the motivation, goals and non-goals of
+this KEP.  Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+### Goals
+
+- provide a node-level abstraction of the cluster-wide `Service` semantics through an API
+- allow easier, more stable proxy implementations that don't need updates when `Service` business
+  logic changes
+- provide a client library with minimal dependencies
+- include equivalent implementations of in-project ones (userland, iptables and ipvs)
+- create reusable library elements for some proxy logic, such as conntrack cleaning, which might be called from different proxy implementations
+
+### Non-Goals
+
+- provide equivalent implementations of all third-party proxiers
+- force new users to run kpng in two separate daemons - this has performance benefits and is optional
```

Comment:
You introduce this as a non-goal without even having introduced it as a _concept_ yet. Non-goals is probably not the right place to discuss this tradeoff.

(Also, this isn't about forcing "users" to do anything, it's about forcing/not forcing _implementors_ to write their proxy implementations in a specific way.)

(Also also, the statement as written is totally unclear which way you are saying has performance benefits.)

(Also also also, if one way has performance benefits then why would anyone want to do it the other way? IIRC in the original proposal, there was the idea that you could run a single "core logic" daemon _per cluster_, and each node's "backend implementation" would talk to the same "core logic" daemon, so you only had a single daemon processing EndpointSlice updates, instead of having every node have to do it. But I don't see anything in the current KEP that suggests this is still possible, and given that, I don't really see any good reason for having two daemons, other than the 100% dubious "don't need to rebuild" thing...)

Reply:
rewrote it and pushing shortly, also removed words like forced etc...